README.md
# OCSP-RS

[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
[![codecov](https://codecov.io/gh/maicallist/ocsp-rs/branch/master/graph/badge.svg?token=TMNVADDBOK)](https://codecov.io/gh/maicallist/ocsp-rs)
![master](https://github.com/maicallist/ocsp-rs/actions/workflows/master.yml/badge.svg?branch=master)

## ocsp-rs supports encoding/decoding OCSP request and response

## Features

- request encoding [WIP]
- request decoding
- response encoding
- response decoding [WIP]

## Usage

```toml
[dependencies]
ocsp = "0.4"
```

### Server side

#### 1. Parsing OCSP request

```rust
use ocsp::request::OcspRequest;

let recv_request: BytesMut = BytesMut::new();

// reading http payload to `recv_request`

let ocsp_request = OcspRequest::parse(&recv_request[..]).unwrap();

// get CertId from request
let cid_list = ocsp_request.extract_certid_owned();
```

#### 2. Generating OCSP response

```rust
use ocsp::{
    common::asn1::{CertId, GeneralizedTime, Oid},
    oid::{ALGO_SHA256_WITH_RSA_ENCRYPTION_DOT, OCSP_RESPONSE_BASIC_DOT},
    response::{
        BasicResponse, CertStatus as OcspCertStatus, CertStatus, CertStatusCode, CrlReason,
        OcspRespStatus, OcspResponse, OneResp, ResponderId, ResponseBytes, ResponseData,
        RevokedInfo,
    },
};


let key = [0x36, 0x6f, 0x35, 0xfb, 0xef, 0x16, 0xc6, 0xba, 0x8a, 0x31, 0x83, 0x42, 0x6d, 0x97, 0xba, 0x89, 0x4d, 0x55, 0x6e, 0x91];
let id = ResponderId::new_key_hash(&key); // responding by id

// year, month, day, hour(24), minute, second
let produce = GeneralizedTime::new(2021, 1, 12, 21, 26, 43).unwrap();
// you can extract cid from request
let oid = Oid::new_from_dot("1.3.14.3.2.26").unwrap();
let name = vec![ 0x69, 0x4d, 0x18, 0xa9, 0xbe, 0x42, 0xf7, 0x80, 0x26, 0x14, 0xd4, 0x84, 0x4f, 0x23, 0x60, 0x14, 0x78, 0xb7, 0x88, 0x20];
let key = vec![ 0x39, 0x7b, 0xe0, 0x02, 0xa2, 0xf5, 0x71, 0xfd, 0x80, 0xdc, 0xeb, 0x52, 0xa1, 0x7a, 0x7f, 0x8b, 0x63, 0x2b, 0xe7, 0x55];
let sn = vec![0x41, 0x30, 0x09, 0x83, 0x33, 0x1f, 0x9d, 0x4f];
let certid = CertId::new(oid.clone(), &name, &key, &sn);

let good = OcspCertStatus::new(CertStatusCode::Good, None);
let gt = GeneralizedTime::new(2021, 1, 12, 3, 26, 43).unwrap();

let one = OneResp {
    cid: certid.clone(),
    cert_status: good,
    this_update: gt,
    next_update: None,
    one_resp_ext: None,
};

let sn2 = vec![0x63, 0x78, 0xe5, 0x1d, 0x44, 0x8f, 0xf4, 0x6d];
let certid2 = CertId::new(oid, &name, &key, &sn2);
let rev_t = GeneralizedTime::new(2020, 11, 30, 1, 48, 25).unwrap();
let rev_info = RevokedInfo::new(rev_t, Some(CrlReason::OcspRevokeUnspecified));
let revoke = OcspCertStatus::new(CertStatusCode::Revoked, Some(rev_info));
let two = OneResp {
    cid: certid2,
    cert_status: revoke,
    this_update: gt,
    next_update: None,
    one_resp_ext: None,
};

let list = [one, two].to_vec();
let data = ResponseData::new(id, produce, list, None);
// other signatures also supported, see oid
// equivalent to
// let oid = Oid::new_from_dot("1.2.840.113549.1.1.5").await.unwrap();
let oid = Oid::new_from_dot(ALGO_SHA256_WITH_RSA_ENCRYPTION_DOT).unwrap();

let some_signing_machine = || async { vec![ 0x00 ] };
let sign = some_signing_machine().await; //example signature

let basic = BasicResponse::new(data, oid, sign, None);
// equivalent to
// let resp_type = Oid::new_from_dot("1.3.6.1.5.5.7.48.1.1").await.unwrap();
let resp_type = Oid::new_from_dot(OCSP_RESPONSE_BASIC_DOT).unwrap();
let bytes = ResponseBytes::new_basic(resp_type, basic).unwrap();
let ocsp = OcspResponse::new_success(bytes);
let resp_binary = ocsp.to_der().unwrap();

// return resp_binary as response body
```

---

### Client side [WIP]


CHANGELOG
# OCSP-RS Changelog

## [0.4.0] - 2023.1.9

* [BREAKING CHANGE] remove unnecessary async keyword

## [0.3.3] - 2021.8.27

* add metadata for crate.io release

## [0.3.2] - 2021.6.17

* rework common module logging missed in last release

## [0.3.1] - 2021.6.17

* rework logging messages

## [0.3.0] - 2021.5.27

* allowing extracting owned cids from request

## [0.2.0] - 2021.5.27

* [BREAKING CHANGE] rename one_resp to cid to eliminate confusion

## [0.1.1] - 2021.5.26

* allowing CertId to be HashMap key


src/tbd.rs
//! common functions for ocsp request and response

use super::err::OcspError;
use asn1_der::typed::{DerDecodable, Sequence};
use futures::future::{BoxFuture, FutureExt};
use hex;

use super::common::{TryIntoSequence, CERTID_TAG};

/// OCSP request binary object
///
///```rust
/// use ocsp_rs::asn1_common::*;
/// use asn1_der::typed::DerDecodable;
///
/// let ocsp_req = "306e306c304530433041300906052b0e\
///03021a05000414694d18a9be42f78026\
///14d4844f23601478b788200414397be0\
///02a2f571fd80dceb52a17a7f8b632be7\
///5502086378e51d448ff46da223302130\
///1f06092b060105050730010204120410\
///1cfc8fa3f5e15ed760707bc46670559b";
/// let ocsp_bin = hex::decode(ocsp_req).unwrap();
/// let asn1 = asn1_der::DerObject::decode(&ocsp_bin[..]).unwrap();
/// println!("asn1 tag: {:02X}, asn1 header: {:02X?}, asn1 value: {:02X?}", asn1.tag(), asn1.header(), asn1.value());
/// let seq = asn1_der::typed::Sequence::decode(asn1.raw()).unwrap();
/// let req = OcspAsn1Der::parse(&asn1).unwrap();
///```
/// above binary data has the following structure:
///
/// <table>
///   <tr>
///     <th>ASN1 hex</th>
///     <th>ASN1 scheme</th>
///   </tr>
///   <tr>
///     <td>
///       <pre>
/// 30 6E               %% RFC OCSPRequest
/// | 30 6C             %% RFC TBSRequest
/// |   30 45           %% RFC requestList
/// |   | 30 43         %% RFC Request
/// |   |   30 41       %% RFC CertID
/// |   |     30 09
/// |   |       06 05 2B0E03021A  --- OID
/// |   |       05 00             --- NULL
/// |   |     04 14 694D18A9BE42F7802614D4844F23601478B78820  --- OCTET
/// |   |     04 14 397BE002A2F571FD80DCEB52A17A7F8B632BE755  --- OCTET
/// |   |     02 08 6378E51D448FF46D  --- INT
///         30 imagine    %% RFC singleRequestExtensions
/// |   A2 23  --- EXPLICIT TAG 2   %% RFC requestExtensions
/// |   | 30 21
/// |   |   30 1F
/// |   |     06 09 2B0601050507300102
/// |   |     04 12 04101CFC8FA3F5E15ED760707BC46670559B
/// |
/// |--- Sequence(30), 110 bytes(6E)
///       </pre>
///     </td>
///     <td>
///       <pre>
/// SEQUENCE {
///   SEQUENCE {
///   | SEQUENCE {
///   |   SEQUENCE {
///   |     SEQUENCE {
///   |     | SEQUENCE {
///   |     |    OBJECTIDENTIFIER 1.3.14.3.2.26 (id_sha1)
///   |     |    NULL
///   |     | }
///   |     | OCTETSTRING 694d18a9be42f7802614d4844f23601478b78820
///   |     | OCTETSTRING 397be002a2f571fd80dceb52a17a7f8b632be755
///   |     | INTEGER 0x6378e51d448ff46d
///   |     }
///   |   }
///   | }
///   | [2] {
///   |   SEQUENCE {
///   |   | SEQUENCE {
///   |   |   OBJECTIDENTIFIER 1.3.6.1.5.5.7.48.1.2
///   |   |   OCTETSTRING 04101cfc8fa3f5e15ed760707bc46670559b
///   |   | }
///   |   }
///   | }
///   }
/// }
///       </pre>
///     </td>
/// </table>
///
/// Or Verifying two certificates in one request
/// ```asn1
/// 30 81B5(181)
///   30 81B2(178)
///     30 818A(138)
///     | 30 43
///     |   30 41
///     |     30 09
///     |       06 05 2B0E03021A
///     |       05 00
///     |     04 14 694D18A9BE42F7802614D4844F23601478B78820
///     |     04 14 397BE002A2F571FD80DCEB52A17A7F8B632BE755
///     |     02 08 41300983331F9D4F
///     | 30 43
///     |   30 41
///     |     30 09
///     |       06 05 2B0E03021A
///     |       05 00
///     |     04 14 694D18A9BE42F7802614D4844F23601478B78820
///     |     04 14 397BE002A2F571FD80DCEB52A17A7F8B632BE755
///     |     02 08 6378E51D448FF46D
///     A2 23
///     | 30 21
///     |   30 1F
///     |     06 09 2B0601050507300102
///     |     04 12 04105E7A74E51C861A3F79454658BB090244
/// ```
///
#[allow(dead_code)]
pub struct OcspAsn1Der<'d> {
    seq: Sequence<'d>,
}

#[allow(dead_code)]
impl<'d> OcspAsn1Der<'d> {
    /// create Sequence type from raw der
    pub fn parse<T: TryIntoSequence<'d>>(t: &'d T) -> Result<Self, OcspError>
    where
        T: 'd + TryIntoSequence<'d, Error = OcspError>,
    {
        t.try_into().map(|s| OcspAsn1Der { seq: s })
    }

    /// Extracting CertId Sequence from ASN1 DER data.  
    /// tags must match following hex order:  
    /// 30(6, 5), 4, 4, 2  
    ///
    /// - **self.seq** A sequence to be examined
    /// - **tag** CERTID tag array  
    /// per rfc 6960 CERTID matches sequence of OID, OCTET, OCTET, INTEGER,  
    /// thus tag should contain 0x06, 0x05, 0x04, 0x04, 0x02 as result.  
    /// In practice, openssl has 0x05 after OID 0x06.  
    /// - **value** corresponding value of @tag array  
    ///
    /// extracted CERTID tag and value stores in 'tag' and 'val'
    fn extract_certid_raw(
        &'d self,
        tag: &'d mut Bytes,
        val: &'d mut Vec<Bytes>,
    ) -> BoxFuture<'d, Result<u8, OcspError>> {
        async move {
            let mut examine = false;
            for i in 0..self.seq.len() {
                let tmp = self.seq.get(i).map_err(OcspError::Asn1DecodingError)?;
                match tmp.tag() {
                    0x30 => {
                        let mut raw = tmp.header().to_vec();
                        raw.extend(tmp.value());
                        let tseq =
                            Sequence::decode(&raw[..]).map_err(OcspError::Asn1DecodingError)?;

                        match OcspAsn1Der::extract_certid_raw(&OcspAsn1Der { seq: tseq }, tag, val)
                            .await?
                        {
                            0 => break,
                            1 => continue,
                            _ => return Err(OcspError::Asn1ExtractionUnknownError),
                        }
                    }
                    0x02 | 0x04 | 0x05 | 0x06 => {
                        tag.push(tmp.tag());
                        val.push(tmp.value().to_vec());
                        examine = true;
                    }
                    _ => break,
                }
            }

            if examine {
                match count_matching_tags(&CERTID_TAG, tag).await % CERTID_TAG.len() {
                    0 | 2 => return Ok(1),
                    _ => {
                        tag.truncate(tag.len() / CERTID_TAG.len());
                        val.truncate(val.len() / CERTID_TAG.len());
                        return Ok(0);
                    }
                }
            }

            Ok(1)
        }
        .boxed()
    }

    /// extracting certificate serial number in hex from OCSP request/response
    pub async fn extract_cert_sn(self) -> Result<Vec<String>, OcspError> {
        let mut tag = Vec::new();
        let mut val = Vec::new();
        match OcspAsn1Der::extract_certid_raw(&self, &mut tag, &mut val).await? {
            0 => return Err(OcspError::Asn1MismatchError),
            1 => {}
            _ => return Err(OcspError::Asn1ExtractionUnknownError),
        };

        // NOTICE: wait for async hex
        let sn_hex_list: Vec<String> = async move {
            val.into_iter()
                .skip(4)
                .step_by(5)
                .map(|v| hex::encode(v))
                .collect()
        }
        .await;

        Ok(sn_hex_list)
    }
}

/// count number of matching tag to a sequence
/// - **target** target tag sequence
/// - **tbm** tag sequence to be examined
/// - **@return** number of tags matching target
async fn count_matching_tags(target: &[u8], tbm: &[u8]) -> usize {
    let mut tt = target.to_vec();
    while tt.len() < tbm.len() {
        tt.extend(target);
    }
    let tt = &tt[..tbm.len()];
    tbm.iter().zip(tt).filter(|(m, t)| m == t).count()
}

#[cfg(test)]
mod test {
    use super::*;
    use asn1_der::DerObject;
    use tokio;

    /// extracting single certid
    #[tokio::test]
    async fn ocsp_req_get_certid_raw() {
        let ocsp_req_hex = "306e306c304530433041300906052b0e\
    03021a05000414694d18a9be42f78026\
    14d4844f23601478b788200414397be0\
    02a2f571fd80dceb52a17a7f8b632be7\
    5502086378e51d448ff46da223302130\
    1f06092b060105050730010204120410\
    1cfc8fa3f5e15ed760707bc46670559b";
        let ocsp_req = hex::decode(ocsp_req_hex).unwrap();
        let seq = Sequence::decode(&ocsp_req[..]).unwrap();
        let der = OcspAsn1Der { seq: seq };
        let mut tag = Vec::new();
        let mut val = Vec::new();
        let _ = der.extract_certid_raw(&mut tag, &mut val).await;
        //println!(
        //    "-----tag-----\n{:02X?}\n{:02X?}\n------end of line -----",
        //    tag, val
        //);
        assert_eq!(tag, vec![0x06u8, 0x05, 0x04, 0x04, 0x02]);
        assert_eq!(
            val,
            vec![
                vec![0x2b, 0x0e, 0x03, 0x02, 0x1a],
                vec![],
                vec![
                    0x69, 0x4d, 0x18, 0xa9, 0xbe, 0x42, 0xf7, 0x80, 0x26, 0x14, 0xd4, 0x84, 0x4f,
                    0x23, 0x60, 0x14, 0x78, 0xb7, 0x88, 0x20
                ],
                vec![
                    0x39, 0x7b, 0xe0, 0x02, 0xa2, 0xf5, 0x71, 0xfd, 0x80, 0xdc, 0xeb, 0x52, 0xa1,
                    0x7a, 0x7f, 0x8b, 0x63, 0x2b, 0xe7, 0x55
                ],
                vec![0x63, 0x78, 0xe5, 0x1d, 0x44, 0x8f, 0xf4, 0x6d]
            ]
        );
    }

    #[tokio::test]
    // sequence from parse method
    async fn ocsp_sequence_parse_derobj() {
        let ocsp_req_hex = "306e306c304530433041300906052b0e\
    03021a05000414694d18a9be42f78026\
    14d4844f23601478b788200414397be0\
    02a2f571fd80dceb52a17a7f8b632be7\
    5502086378e51d448ff46da223302130\
    1f06092b060105050730010204120410\
    1cfc8fa3f5e15ed760707bc46670559b";
        let ocsp_req = hex::decode(ocsp_req_hex).unwrap();
        let derobj = DerObject::decode(&ocsp_req[..]).unwrap();
        let der = OcspAsn1Der::parse(&derobj).unwrap();
        let mut tag = Vec::new();
        let mut val = Vec::new();
        let _ = der.extract_certid_raw(&mut tag, &mut val).await;
        assert_eq!(tag, vec![0x06u8, 0x05, 0x04, 0x04, 0x02]);
        assert_eq!(
            val,
            vec![
                vec![0x2b, 0x0e, 0x03, 0x02, 0x1a],
                vec![],
                vec![
                    0x69, 0x4d, 0x18, 0xa9, 0xbe, 0x42, 0xf7, 0x80, 0x26, 0x14, 0xd4, 0x84, 0x4f,
                    0x23, 0x60, 0x14, 0x78, 0xb7, 0x88, 0x20
                ],
                vec![
                    0x39, 0x7b, 0xe0, 0x02, 0xa2, 0xf5, 0x71, 0xfd, 0x80, 0xdc, 0xeb, 0x52, 0xa1,
                    0x7a, 0x7f, 0x8b, 0x63, 0x2b, 0xe7, 0x55
                ],
                vec![0x63, 0x78, 0xe5, 0x1d, 0x44, 0x8f, 0xf4, 0x6d]
            ]
        );
    }

    #[tokio::test]
    // sequence from parse method
    async fn ocsp_sequence_parse_vec() {
        let ocsp_req_hex = "306e306c304530433041300906052b0e\
    03021a05000414694d18a9be42f78026\
    14d4844f23601478b788200414397be0\
    02a2f571fd80dceb52a17a7f8b632be7\
    5502086378e51d448ff46da223302130\
    1f06092b060105050730010204120410\
    1cfc8fa3f5e15ed760707bc46670559b";
        let ocsp_req = hex::decode(ocsp_req_hex).unwrap();
        let der = OcspAsn1Der::parse(&ocsp_req).unwrap();
        let mut tag = Vec::new();
        let mut val = Vec::new();
        let _ = der.extract_certid_raw(&mut tag, &mut val).await;
        assert_eq!(tag, vec![0x06u8, 0x05, 0x04, 0x04, 0x02]);
        assert_eq!(
            val,
            vec![
                vec![0x2b, 0x0e, 0x03, 0x02, 0x1a],
                vec![],
                vec![
                    0x69, 0x4d, 0x18, 0xa9, 0xbe, 0x42, 0xf7, 0x80, 0x26, 0x14, 0xd4, 0x84, 0x4f,
                    0x23, 0x60, 0x14, 0x78, 0xb7, 0x88, 0x20
                ],
                vec![
                    0x39, 0x7b, 0xe0, 0x02, 0xa2, 0xf5, 0x71, 0xfd, 0x80, 0xdc, 0xeb, 0x52, 0xa1,
                    0x7a, 0x7f, 0x8b, 0x63, 0x2b, 0xe7, 0x55
                ],
                vec![0x63, 0x78, 0xe5, 0x1d, 0x44, 0x8f, 0xf4, 0x6d]
            ]
        );
    }

    #[tokio::test]
    // get two certid from request
    async fn ocsp_req_multiple_certid_raw() {
        let ocsp_req_hex = "3081b53081b230818a30433041300906\
052b0e03021a05000414694d18a9be42\
f7802614d4844f23601478b788200414\
397be002a2f571fd80dceb52a17a7f8b\
632be755020841300983331f9d4f3043\
3041300906052b0e03021a0500041469\
4d18a9be42f7802614d4844f23601478\
b788200414397be002a2f571fd80dceb\
52a17a7f8b632be75502086378e51d44\
8ff46da2233021301f06092b06010505\
07300102041204105e7a74e51c861a3f\
79454658bb090244";
        let ocsp_req = hex::decode(ocsp_req_hex).unwrap();
        let seq = Sequence::decode(&ocsp_req[..]).unwrap();
        let der = OcspAsn1Der { seq: seq };
        let mut tag = Vec::new();
        let mut val = Vec::new();
        let _ = der.extract_certid_raw(&mut tag, &mut val).await;
        assert_eq!(
            tag,
            vec![0x06u8, 0x05, 0x04, 0x04, 0x02, 0x06, 0x05, 0x04, 0x04, 0x02]
        );
        assert_eq!(
            val,
            vec![
                vec![0x2b, 0x0e, 0x03, 0x02, 0x1a],
                vec![],
                vec![
                    0x69, 0x4d, 0x18, 0xa9, 0xbe, 0x42, 0xf7, 0x80, 0x26, 0x14, 0xd4, 0x84, 0x4f,
                    0x23, 0x60, 0x14, 0x78, 0xb7, 0x88, 0x20
                ],
                vec![
                    0x39, 0x7b, 0xe0, 0x02, 0xa2, 0xf5, 0x71, 0xfd, 0x80, 0xdc, 0xeb, 0x52, 0xa1,
                    0x7a, 0x7f, 0x8b, 0x63, 0x2b, 0xe7, 0x55
                ],
                vec![0x41, 0x30, 0x09, 0x83, 0x33, 0x1f, 0x9d, 0x4f],
                vec![0x2b, 0x0e, 0x03, 0x02, 0x1a],
                vec![],
                vec![
                    0x69, 0x4d, 0x18, 0xa9, 0xbe, 0x42, 0xf7, 0x80, 0x26, 0x14, 0xd4, 0x84, 0x4f,
                    0x23, 0x60, 0x14, 0x78, 0xb7, 0x88, 0x20
                ],
                vec![
                    0x39, 0x7b, 0xe0, 0x02, 0xa2, 0xf5, 0x71, 0xfd, 0x80, 0xdc, 0xeb, 0x52, 0xa1,
                    0x7a, 0x7f, 0x8b, 0x63, 0x2b, 0xe7, 0x55
                ],
                vec![0x63, 0x78, 0xe5, 0x1d, 0x44, 0x8f, 0xf4, 0x6d],
            ]
        );
    }

    #[tokio::test]
    // missing 05 after 06
    async fn ocsp_req_wrong_certid_raw() {
        let ocsp_req_hex = "306c\
306a\
3043\
3041\
303f\
3007\
06052b0e03021a\
0414694d18a9be42f7802614d4844f23601478b78820\
0414397be002a2f571fd80dceb52a17a7f8b632be755\
02086378e51d448ff46d\
a2233021\
301f\
06092b0601050507300102\
041204101cfc8fa3f5e15ed760707bc46670559b";
        let ocsp_req = hex::decode(ocsp_req_hex).unwrap();
        let seq = Sequence::decode(&ocsp_req[..]).unwrap();
        let der = OcspAsn1Der { seq: seq };
        let mut tag = Vec::new();
        let mut val = Vec::new();
        let _ = der.extract_certid_raw(&mut tag, &mut val).await;
        assert_eq!(tag, vec![]);
    }

    #[tokio::test]
    // second sequence mismatch
    // removing first 04 from first certid
    async fn ocsp_req_extract_second_certid_raw() {
        let ocsp_req_hex = "30819e\
30819b\
3074\
302d\
302b\
3009\
06052b0e03021a\
0500\
0414397be002a2f571fd80dceb52a17a7f8b632be755\
020841300983331f9d4f\
3043\
3041\
3009\
06052b0e03021a\
0500\
0414694d18a9be42f7802614d4844f23601478b78820\
0414397be002a2f571fd80dceb52a17a7f8b632be755\
02086378e51d448ff46d\
a2233021301f06092b0601050507300102041204105e7a74e51c861a3f79454658bb090244";
        let ocsp_req = hex::decode(ocsp_req_hex).unwrap();
        let seq = Sequence::decode(&ocsp_req[..]).unwrap();
        let der = OcspAsn1Der { seq: seq };
        let mut tag = Vec::new();
        let mut val = Vec::new();
        let _ = der.extract_certid_raw(&mut tag, &mut val).await;
        assert_eq!(tag, vec![0x06u8, 0x05, 0x04, 0x04, 0x02]);
        assert_eq!(
            val,
            vec![
                vec![0x2b, 0x0e, 0x03, 0x02, 0x1a],
                vec![],
                vec![
                    0x69, 0x4d, 0x18, 0xa9, 0xbe, 0x42, 0xf7, 0x80, 0x26, 0x14, 0xd4, 0x84, 0x4f,
                    0x23, 0x60, 0x14, 0x78, 0xb7, 0x88, 0x20
                ],
                vec![
                    0x39, 0x7b, 0xe0, 0x02, 0xa2, 0xf5, 0x71, 0xfd, 0x80, 0xdc, 0xeb, 0x52, 0xa1,
                    0x7a, 0x7f, 0x8b, 0x63, 0x2b, 0xe7, 0x55
                ],
                vec![0x63, 0x78, 0xe5, 0x1d, 0x44, 0x8f, 0xf4, 0x6d]
            ]
        );
    }

    #[tokio::test]
    // extract two sn from request
    async fn ocsp_req_get_cert_sn() {
        let ocsp_req_hex = "3081b53081b230818a30433041300906\
052b0e03021a05000414694d18a9be42\
f7802614d4844f23601478b788200414\
397be002a2f571fd80dceb52a17a7f8b\
632be755020841300983331f9d4f3043\
3041300906052b0e03021a0500041469\
4d18a9be42f7802614d4844f23601478\
b788200414397be002a2f571fd80dceb\
52a17a7f8b632be75502086378e51d44\
8ff46da2233021301f06092b06010505\
07300102041204105e7a74e51c861a3f\
79454658bb090244";
        let ocsp_req = hex::decode(ocsp_req_hex).unwrap();
        let seq = Sequence::decode(&ocsp_req[..]).unwrap();
        let der = OcspAsn1Der { seq: seq };
        let v = der.extract_cert_sn().await.unwrap();
        assert_eq!(v, vec!["41300983331f9d4f", "6378e51d448ff46d"])
    }
}
